### PR TITLE
feat(storage): Gen2 Storage Download/Upload

### PR DESCRIPTION
**Pull request by tylerjroach**
- [ ] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*

*How did you test these changes?*
(Please add a line here how the changes were tested)

*Documentation update required?*
- [ ] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [ ] Added Unit Tests
- [ ] Added Integration Tests
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StoragePathDownloadTest.kt [hunk 1]
```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (sktimalsina) [inline aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StoragePathDownloadTest.kt:2]: Nit: 2024

Comment (tylerjroach) [reply aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StoragePathDownloadTest.kt:2]: I need to go through and check all my licenses. This one is a copy/paste issue, but I believe my auto license header was disabled and not applied to some of the new files. Will make a note.

**Comment by tylerjroach** [comment]
Looks like I didn't include upload file/inputstream integration tests on this PR. Will add to follow up

File: aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StoragePathUploadFileOperation.kt [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+import com.amplifyframework.storage.s3.transfer.TransferListener
+import com.amplifyframework.storage.s3.transfer.TransferObserver
+import java.io.File
+import java.util.UUID
+import java.util.concurrent.ExecutorService
+
+/**
+ * An operation to upload a file from AWS S3.
+ */
+internal class AWSS3StoragePathUploadFileOperation @JvmOverloads internal constructor(
```
Comment (mattcreaser) [inline aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StoragePathUploadFileOperation.kt:42]: Do we need `@JvmOverloads` on these operation constructors? Seems like the plugin is always supplying all values.

Comment (tylerjroach) [reply aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StoragePathUploadFileOperation.kt:42]: Will remove. This was from copy paste

Comment (tylerjroach) [reply aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StoragePathUploadFileOperation.kt:42]: Will remove on InputStream operation as well.

File: core/src/main/java/com/amplifyframework/storage/StoragePath.kt [hunk 3]
```diff
@@ -2,26 +2,53 @@ package com.amplifyframework.storage
...
         @JvmStatic
         fun fromIdentityId(identityIdPathResolver: IdentityIdPathResolver): StoragePath =
             IdentityIdProvidedStoragePath(identityIdPathResolver)
     }
 }
 
+/**
+ * StoragePath that was created with the full String path.
+ */
 data class StringStoragePath internal constructor(private val path: String) : StoragePath() {
```
[the remark this answers is not in the document]
Comment (tylerjroach) [reply core/src/main/java/com/amplifyframework/storage/StoragePath.kt:43]: We talked on api review and there could be some valid reason to check type that we decided it was valid to expose.

**Comment by mattcreaser** [comment]
General comment: the `*Operation` classes have a lot of duplicated logic. I know that's how the existing classes are already written but we could certainly reduce that by using composition patterns in a future refactor.

File: aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/extensions/StoragePath.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import kotlinx.coroutines.runBlocking
+
+internal fun StoragePath.toS3ServiceKey(authCredentialsProvider: AuthCredentialsProvider): String {
+    val stringPath = when (this) {
+        is StringStoragePath -> {
+            resolvePath()
+        }
+        is IdentityIdProvidedStoragePath -> {
+            val identityId = try {
+                runBlocking {
```
[the remark this answers is not in the document]
Comment (tylerjroach) [reply aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/extensions/StoragePath.kt:18]: Will make this change in follow up pr

**Review by mattcreaser — APPROVED** [review]

**Pull request merged**